nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/lifecycle/event/BulkLifeCycleChangeListener.java

```diff
@@ -195,15 +195,19 @@ protected void changeChildrenState(CoreSession session, String transition, Strin
      */
     protected void changeDocumentsState(CoreSession session, String transition, String targetState,
             DocumentModelList docs) {
+        TrashService trashService = Framework.getService(TrashService.class);
         for (DocumentModel doc : docs) {
+            if (LifeCycleConstants.UNDELETE_TRANSITION.equals(transition)
+                    && trashService.isMangledName(doc.getName())) {
+                continue;
```

Comment:
Not sure it's what we want, we do not follow undelete transition anymore if the document has a mangled name?
This breaks the recursiveness for the undelete transition and so it may break the lifecycle trash service when it's on. Could you add a test for that? I think we need at least two levels:
> fold1
|_ fold2
&nbsp;&nbsp;&nbsp;|_ doc

Scenario could be something close to:
- trash _fold2_ -> _fold2_ is now mangled and trashed, doc is trashed
- trash _fold1_ -> _fold1_ is now mangled and trashed
- untrash _fold1_ -> _fold1_ is now unmangled and untrashed, so do _fold2_ and _doc_

I suspect that _doc_ is not untrashed anymore with `LifeCycleTrashService`.

Reply:
Yes this is worth testing. I'm on it !
I did this based on the following:
This bulk listener is an event handler that always comes on the "second wave" after a parent has been deleted / undeleted.
When undeleting a parent of an explicitly trashed we want this not to be propagated.
So it's all based on "this doesn't handle the explicit untrashing of a document but only children"

Reply:
Indeed, we may have the good behavior currently and my scenario is not correct regarding behavior we want to introduce. Correct me if I'm wonrg, in the context above, _fold2_ is not untrashed right? and so does _doc_.

Reply:
Yes, they remain in the trashbin as they were thrown away independently from fold1.

Reply:
It happens the `PropertyTrashService` breaks because the query to filter explicitly trashed items doesn't filter their children which are not explicitly trashed.
The actual query gets all the children with the parent in their parent ids, then it only filters those with a mangled name but not their children...

Reply:
I think we need an api to check on the path of a document, then if it contains a trashed name, we know one of the ancestors is trashed and drop. That would prevent us from recursively climb back up the hierarchy to check

Reply:
It becomes a bit complex plus it will probably decrease performance of bulk action because for each items we will check its parent for a mangled name and if not it will check its grand parent, etc. (a simulation can be created to check that)

Furthermore, the whole logic leverages the mangled name whereas it is not reliable for all cases. A user can move the document and change the name.

I would be in favor to pause this development and check with PO if we still want this behavior regarding the technical complexity.